pyro/infer/traceenum_elbo.py

```diff
@@ -20,6 +19,132 @@
+            if dim in dims and term.shape[dim] > 1:
+                neighbors.setdefault(term, []).append(dim)
+                neighbors.setdefault(dim, []).append(term)
+        if term not in neighbors:
+            yield [term], set()
+
+    # Partition the bipartite graph into connected components for contraction.
+    while neighbors:
+        v, pending = neighbors.popitem()
+        component = OrderedDict([(v, None)])  # used as an OrderedSet
+        for v in pending:
+            component[v] = None
+        while pending:
+            v = pending.pop()
+            for v in neighbors.pop(v):
+                if v not in component:
+                    component[v] = None
+                    pending.append(v)
+
+        # Split this connected component into tensors and dims.
+        component_terms = [v for v in component if isinstance(v, torch.Tensor)]
+        component_dims = set(v for v in component if not isinstance(v, torch.Tensor))
+        yield component_terms, component_dims
+
+
+def _contract(log_factors, enum_boundary):
+    """
+    Contract out all enumeration dims in a tree of log_factors in-place, via
+    message passing. This function should be deterministic.
+
+    :param OrderedDict log_factors: a dictionary mapping ordinals to lists of
+        tensors. An ordinal is a frozenset of ``CondIndepStack`` frames.
+    :param int enum_boundary: a dimension (counting from the right) such that
+        all dimensions left of this dimension are enumeration dimensions.
+    """
+    # First close the set of ordinals under intersection (greatest lower bound),
+    # ensuring that the ordinals are arranged in a tree structure.
+    pending = list(log_factors)
+    while pending:
+        t = pending.pop()
+        for u in list(log_factors):
+            tu = t & u
+            if tu not in log_factors:
+                log_factors[tu] = []
+                pending.append(tu)
+
+    # Collect all enumeration dimensions.
+    enum_dims = defaultdict(set)
+    for t, terms in log_factors.items():
+        enum_dims[t] |= set(i for term in terms
+                            for i in range(-term.dim(), enum_boundary)
+                            if term.shape[i] > 1)
+
+    # Recursively combine terms in different iarange contexts.
+    while len(log_factors) > 1 or any(enum_dims.values()):
+        leaf = max(log_factors, key=len)
+        leaf_terms = log_factors.pop(leaf)
+        leaf_dims = enum_dims.pop(leaf)
+        remaining_dims = set.union(set(), *enum_dims.values())
+        contract_dims = leaf_dims - remaining_dims
+        contract_frames = (frozenset.intersection(*(leaf - t for t in log_factors if t < leaf))
+                           if log_factors else frozenset())
+        parent = leaf - contract_frames
+        enum_dims[parent] |= leaf_dims & remaining_dims
+        log_factors.setdefault(parent, [])
+        for terms, dims in _partition_terms(leaf_terms, contract_dims):
```

Comment:
@neerajprad this differs from `TraceLogProbEvaluator` in that it now splits contraction terms into connected components wrt the shared-enum-dim graph. This is needed in examples where we need to delay the `sumproduct` (and its resulting broadcast) until the parent ordinal.

See the HMM example for motivation: the observations `y` are in an iarange that we need to contract *before* combining with the `x`s outside of the `iarange`. If we put all of the `y`s in a single `sumproduct`, then we would need to preserve all the dimensions for the next ordinal up. In this PR we instead contract the `y`s separately and move each of them to the parent ordinal.